src/aggregate/aggregate_request.c

```diff
@@ -1210,6 +1210,45 @@ int buildOutputPipeline(AREQ *req, QueryError *status) {
   return REDISMODULE_ERR;
 }
 
+static void PushBefore(ResultProcessor *rp_to_place, ResultProcessor *upstream) {
+  rp_to_place->upstream = upstream->upstream;
+  upstream->upstream = rp_to_place;
+  rp_to_place->parent = upstream->parent;
+}
+
+int SafeRedisKeyspaceAccessPipeline(AREQ *req, ResultProcessor *first_to_access_redis,

```

Comment:
I would call it "new buffer-locker and unlocker RP" in some way, and explain the safe keyspace access in a comment

Reply:
it is based on results_processor_job_pipline format